pkg/vault/contracts/authorizer/TimelockAuthorizerMigrator.sol

```diff
@@ -61,27 +70,41 @@ contract TimelockAuthorizerMigrator {
         vault = _vault;
 
         for (uint256 i = 0; i < _rolesData.length; i++) {
-            rolesData.push(OldRoleData(_rolesData[i].role, _rolesData[i].target));
+            RoleData memory roleData = _rolesData[i];
+            // We require that any permissions being copied from the old Authorizer must exist on the old Authorizer.
+            // This simplifies verification of the permissions being added to the new TimelockAuthorizer.
+            require(_oldAuthorizer.canPerform(roleData.role, roleData.grantee, roleData.target), "UNEXPECTED_ROLE");
+            rolesData.push(roleData);
+        }
+        for (uint256 i = 0; i < _grantersData.length; i++) {
+            // There's no concept of a "granter" on the old Authorizer so we cannot verify these onchain.
+            // We must manually verify that these permissions are set sensibly.
+            grantersData.push(RoleData(_grantersData[i].grantee, _grantersData[i].role, _grantersData[i].target));
+        }
+        for (uint256 i = 0; i < _revokersData.length; i++) {
+            // Similarly to granters, we must manually verify that these permissions are set sensibly.
+            revokersData.push(RoleData(_revokersData[i].grantee, _revokersData[i].role, _revokersData[i].target));
         }
 
         // Enqueue a root change execution in the new authorizer to set it to the desired root address
         rootChangeExecutionId = _newAuthorizer.scheduleRootChange(_root, _arr(address(this)));
     }
 
     /**
-     * @dev Tells whether the migration has been completed or not
+     * @notice Returns whether the migration has been completed or not
      */
     function isComplete() public view returns (bool) {
-        return migratedRoles >= rolesData.length;
+        // As we set up the revoker roles last we can use them to determine whether the full migration is complete.
+        return revokersMigrated >= revokersData.length;
```

Comment:
Seems a bit hacky to check this way. It depends on the implementation in _migrate(). Could use a boolean flag for this, and set it at the end of _migrate(). Or at least document in _migrate() that there is an order dependency, for future proofing. If somebody ever added something after, or changed the order, this could break

Reply:
I've addressed this in [810869a](https://github.com/balancer-labs/balancer-v2-monorepo/pull/1325/commits/810869ac454bfb70a156752b75452b739fa6f9e8)